File: 3006-find-beautiful-indices-in-the-given-array-i/3006-find-beautiful-indices-in-the-given-array-i.py
class Solution:
    def beautifulIndices(self, s: str, a: str, b: str, k: int) -> List[int]:
        def find_all(x, y):
            res = []
            i = x.find(y)
            while i != -1:
                res.append(i)
                i = x.find(y, i + 1)
            return res
        ia = find_all(s, a)
        ib = find_all(s, b)
        if not ia or not ib:
            return []
        res = []
        iib = 0
        for i in ia:
            while iib < len(ib) and i - ib[iib] > k:
                iib += 1
            if iib < len(ib) and abs(i - ib[iib]) <= k:
                res.append(i)
        return res

File: 3006-find-beautiful-indices-in-the-given-array-i/3006-find-beautiful-indices-in-the-given-array-i.java
class Solution {
    public List<Integer> beautifulIndices(String s, String a, String b, int k) {
        List<Integer> res = new ArrayList<>();
        int i = s.indexOf(a);
        int j = s.indexOf(b);
        while (i != -1 && j != -1) {
            if (Math.abs(i - j) <= k) {
                res.add(i);
                i = s.indexOf(a, i + 1);
            } else if (i > j) {
                j = s.indexOf(b, i - k);
            } else {
                i = s.indexOf(a, j - k);
            }
        }
        return res;
    }
}